### DISCUSSION TITLE
Is it possible to merge events with event transformer?

### DISCUSSION THREAD

**Discussion by Jebbyo** [MMRIZE/MMM-CalendarExt3#188] [General]
My wife takes appointments for her business, so she has a lot of events each day she works. I want to be able to merge back-to-back events and show the start time of the first event and the end time of the last event to reduce the number of entries on the calendar. 

Ideally, I would like something that could account for gaps between appointments as well since she will still be at work. Ie) I can set the transformer to combine all events that are <= 2 hours apart, and it will show me the start time of the first event and the end time of the last event. I know it's asking for a lot, and it's probably not even possible, but I wanted to ask.

**Comment by eouia** [comment]
https://forum.magicmirror.builders/topic/16690/mmm-calendarext3/580?_=1733779139269

Instead of `eventTransformer`, we'll use `eventPayload` because, this conversion should be applied to whole events, not a single event.

Insert this property into your config section of CX3 in `config/config.js`
```js
/* `MMM-CalendarExt3` config section in config/config.js */
eventPayload: (payload) => {
	const fulldays = payload.filter(ev => ev.fullDayEvent === true)
	const singleEvents = payload.filter(ev => ev.fullDayEvent === false).sort((a, b) => a.startDate - b.startDate || a.endDate - b.endDate)
	const EventsArranged = singleEvents.reduce((acc, ev) => {
		const dateObj = new Date(+ev.startDate)
		const key = `${ev.title}-${new Date(dateObj).toLocaleDateString('en-CA')}`
		if (!acc.has(key)) {
			acc.set(key, [ev])
		} else {
			acc.get(key).push(ev)
		}
		return acc
	}, new Map())
	const collapsed = Array.from(EventsArranged.entries()).map(([ key, value ]) => {
		const newEvent = { ...value[ 0 ] }
		if (value.length > 1) {
			newEvent.title = newEvent.title + ` <sub class="collapsed">${value.length}</sub>`
			newEvent.endDate = Math.max(...value.map(ev => ev.endDate))
			newEvent.startDate = Math.min(...value.map(ev => ev.startDate))
		}
		return newEvent
	})
	return [ ...fulldays, ...collapsed ]
},
```

And to beautify, add some CSS code into `css/custom.css`
```css
/* css/custom.css */
.CX3 .title sub::after,
#CX3_POPOVER .title sub::after {
  content: "]";
}

.CX3 .title sub::before,
#CX3_POPOVER .title sub::before {
  content: "[";
}
```



---

>-  This code will not affect the fullday events.
>- Not considered every aspect, like inter-day events or untested errors.
